File: SundeeFundee/Sources/SundeeFundeeKit/DomainLayer/Celebration/CelebrationEvent.swift
import Foundation

// MARK: - Celebration Event

/// A celebration-worthy event in the app
public enum CelebrationEvent: Sendable {
    case workoutCompleted(durationSeconds: Int)
    case newPersonalRecord(exerciseName: String, weightKg: Double)
    case programCompleted(programName: String)
    case weightMilestone(exerciseName: String, thresholdKg: Double)
    case newConditioningPR(exerciseName: String, value: Double, scoringType: BenchmarkScoringType)
    case challengeTierCompleted(challengeTitle: String, tierName: String, volumeLbs: Double)
    case challengeCompleted(challengeTitle: String)
}

// MARK: - Formatting Helpers

/// Format a conditioning value based on scoring type
private func formatConditioningValue(_ value: Double, scoringType: BenchmarkScoringType) -> String {
    switch scoringType {
    case .time:
        let totalSeconds = Int(value)
        let minutes = totalSeconds / 60
        let seconds = totalSeconds % 60
        return "\(minutes):\(String(format: "%02d", seconds))"
    case .reps:
        let count = Int(value)
        return count == 1 ? "1 rep" : "\(count) reps"
    case .roundsAndReps:
        let rounds = Int(value) / 10000
        let reps = Int(value) % 10000
        return "\(rounds) rounds + \(reps) reps"
    case .distance:
        return "\(Int(value)) m"
    case .load, .calories:
        return "\(Int(value)) kg"
    }
}

// MARK: - Title & Subtitle

/// Title for a celebration event
public func celebrationTitle(_ event: CelebrationEvent) -> String {
    switch event {
    case .workoutCompleted:
        return "Workout Complete!"
    case .newPersonalRecord:
        return "New Personal Record!"
    case .programCompleted:
        return "Program Complete!"
    case .weightMilestone:
        return "Weight Milestone!"
    case .newConditioningPR:
        return "New Conditioning PR!"
    case .challengeTierCompleted:
        return "Challenge Milestone!"
    case .challengeCompleted:
        return "Challenge Complete!"
    }
}

/// Subtitle for a celebration event
public func celebrationSubtitle(_ event: CelebrationEvent, unit: String = "kg") -> String {
    switch event {
    case .workoutCompleted(let durationSeconds):
        let minutes = durationSeconds / 60
        return minutes > 0 ? "Completed in \(minutes) min" : "Workout saved!"
    case .newPersonalRecord(let exerciseName, let weightKg):
        let converted = unit == "lb" ? weightKg * 2.20462 : weightKg
        let formatted = formatWeight(converted)
        return "\(exerciseName) — \(formatted) \(unit) estimated 1RM"
    case .programCompleted(let programName):
        return "You finished \(programName). Time to level up!"
    case .weightMilestone(let exerciseName, let thresholdKg):
        let converted = unit == "lb" ? thresholdKg * 2.20462 : thresholdKg
        return "\(exerciseName) hit \(Int(converted.rounded())) \(unit)!"
    case .newConditioningPR(let exerciseName, let value, let scoringType):
        let formatted = formatConditioningValue(value, scoringType: scoringType)
        return "\(exerciseName) — \(formatted)"
    case .challengeTierCompleted(let challengeTitle, let tierName, let volumeLbs):
        let formatted = formatWeight(volumeLbs)
        return "\(challengeTitle) — \(tierName) tier at \(formatted) lbs!"
    case .challengeCompleted(let challengeTitle):
        return "You've completed \(challengeTitle)!"
    }
}

/// Format a weight value, removing trailing .0
private func formatWeight(_ value: Double) -> String {
    if value == value.rounded() {
        return "\(Int(value))"
    }
    return String(format: "%.1f", value)
}

File: android/core/src/main/java/com/sundeefundee/core/domain/celebration/CelebrationEvent.kt
package com.sundeefundee.core.domain.celebration

import com.sundeefundee.core.domain.benchmark.BenchmarkScoringType

// MARK: - Celebration Event

/**
 * A celebration-worthy event in the app.
 */
sealed class CelebrationEvent {

    /** A workout was completed. */
    data class WorkoutCompleted(
        val durationSeconds: Int
    ) : CelebrationEvent()

    /** A new personal record was set. */
    data class NewPersonalRecord(
        val exerciseName: String,
        val weightKg: Double
    ) : CelebrationEvent()

    /** A program was completed. */
    data class ProgramCompleted(
        val programName: String
    ) : CelebrationEvent()

    /** A weight milestone was reached. */
    data class WeightMilestone(
        val exerciseName: String,
        val thresholdKg: Double
    ) : CelebrationEvent()

    /** A new conditioning PR was achieved. */
    data class NewConditioningPR(
        val exerciseName: String,
        val value: Double,
        val scoringType: BenchmarkScoringType
    ) : CelebrationEvent()

    /** A challenge tier was completed. */
    data class ChallengeTierCompleted(
        val challengeTitle: String,
        val tierName: String,
        val volumeLbs: Double
    ) : CelebrationEvent()
}

// MARK: - Title & Subtitle

/**
 * Title for a celebration event.
 */
fun celebrationTitle(event: CelebrationEvent): String = when (event) {
    is CelebrationEvent.WorkoutCompleted -> "Workout Complete!"
    is CelebrationEvent.NewPersonalRecord -> "New Personal Record!"
    is CelebrationEvent.ProgramCompleted -> "Program Complete!"
    is CelebrationEvent.WeightMilestone -> "Weight Milestone!"
    is CelebrationEvent.NewConditioningPR -> "New Conditioning PR!"
    is CelebrationEvent.ChallengeTierCompleted -> "Challenge Milestone!"
}

/**
 * Subtitle for a celebration event.
 *
 * @param event The celebration event.
 * @param unit Weight unit string ("kg" or "lb").
 */
fun celebrationSubtitle(event: CelebrationEvent, unit: String = "kg"): String = when (event) {
    is CelebrationEvent.WorkoutCompleted -> {
        val minutes = event.durationSeconds / 60
        if (minutes > 0) "Completed in $minutes min" else "Workout saved!"
    }
    is CelebrationEvent.NewPersonalRecord -> {
        val converted = if (unit == "lb") event.weightKg * 2.20462 else event.weightKg
        val formatted = formatWeight(converted)
        "${event.exerciseName} — $formatted $unit estimated 1RM"
    }
    is CelebrationEvent.ProgramCompleted -> {
        "You finished ${event.programName}. Time to level up!"
    }
    is CelebrationEvent.WeightMilestone -> {
        val converted = if (unit == "lb") event.thresholdKg * 2.20462 else event.thresholdKg
        "${event.exerciseName} hit ${kotlin.math.round(converted).toInt()} $unit!"
    }
    is CelebrationEvent.NewConditioningPR -> {
        val formatted = formatConditioningValue(event.value, event.scoringType)
        "${event.exerciseName} — $formatted"
    }
    is CelebrationEvent.ChallengeTierCompleted -> {
        val formatted = formatWeight(event.volumeLbs)
        "${event.challengeTitle} — ${event.tierName} tier at $formatted lbs!"
    }
}

// MARK: - Private Helpers

/**
 * Format a conditioning value based on scoring type.
 */
private fun formatConditioningValue(value: Double, scoringType: BenchmarkScoringType): String =
    when (scoringType) {
        BenchmarkScoringType.TIME -> {
            val totalSeconds = value.toInt()
            val minutes = totalSeconds / 60
            val seconds = totalSeconds % 60
            "$minutes:${String.format("%02d", seconds)}"
        }
        BenchmarkScoringType.REPS -> {
            val count = value.toInt()
            if (count == 1) "1 rep" else "$count reps"
        }
        BenchmarkScoringType.ROUNDS_AND_REPS -> {
            val rounds = value.toInt() / 10000
            val reps = value.toInt() % 10000
            "$rounds rounds + $reps reps"
        }
        BenchmarkScoringType.DISTANCE -> "${value.toInt()} m"
        BenchmarkScoringType.LOAD -> "${value.toInt()} kg"
        BenchmarkScoringType.CALORIES -> "${value.toInt()} kg"
    }

/**
 * Format a weight value, removing trailing .0.
 */
private fun formatWeight(value: Double): String {
    return if (value == kotlin.math.round(value)) {
        "${value.toInt()}"
    } else {
        String.format("%.1f", value)
    }
}
